### PR TITLE
Adds image layer representation.

### LINKED ISSUE #33: Add a layer representation.
Should consist of `digest`, `size`, and `diffId`.

If the layer is cached, all of these values should be able to be loaded directly.
If the layer is not cached (or needs to be pushed to the registry), it should have another field `content` that is the actual tar.gzip layer blob. In the case that the layer is pulled from a registry, it might not have `diffId` and that would need to be computed by uncompressing the layer blob.

### PR DESCRIPTION
**Pull request by coollog**
Fixes #33

### CODE REVIEW

**Comment by coollog** [comment]
Hold off on review for this. Will be changing changing how this should work.

**Comment by coollog** [comment]
Okay, made this just adding a layer representation.

**Comment by coollog** [comment]
This is good for review now.

File: crepecake/src/main/java/com/google/cloud/tools/crepecake/image/CachedLayer.java [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+ * A {@link Layer} that has been written out (i.e. to a cache) and has its file-backed content BLOB,
+ * digest, size, and diff ID.
+ */
+public class CachedLayer extends Layer {
+
+  private final File file;
+  private final BlobDescriptor blobDescriptor;
+  private final DescriptorDigest diffId;
+
+  public CachedLayer(File file, BlobDescriptor blobDescriptor, DescriptorDigest diffId) {
```
Comment (elharo) [inline crepecake/src/main/java/com/google/cloud/tools/crepecake/image/CachedLayer.java]: diffId? this could use some Javadoc to explain

File: crepecake/src/main/java/com/google/cloud/tools/crepecake/image/CachedLayer.java [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+import com.google.cloud.tools.crepecake.blob.BlobStreams;
+import java.io.File;
+
+/**
+ * A {@link Layer} that has been written out (i.e. to a cache) and has its file-backed content BLOB,
+ * digest, size, and diff ID.
+ */
+public class CachedLayer extends Layer {
+
+  private final File file;
```
Comment (elharo) [inline crepecake/src/main/java/com/google/cloud/tools/crepecake/image/CachedLayer.java]: How are these related? any file with any descriptor and any digest? or do they have to match? If they have to match somehow, will we detect it here if they don't?

File: crepecake/src/main/java/com/google/cloud/tools/crepecake/image/CachedLayer.java [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+    return BlobStreams.from(file);
+  }
+
+  @Override
+  public LayerType getType() {
+    return LayerType.CACHED;
+  }
+
+  @Override
+  public BlobDescriptor getBlobDescriptor() {
```
Comment (elharo) [inline crepecake/src/main/java/com/google/cloud/tools/crepecake/image/CachedLayer.java]: These types are immutable, right?

File: crepecake/src/main/java/com/google/cloud/tools/crepecake/image/Layer.java [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+  public abstract LayerType getType();
+
+  /**
+   * @return the layer's content {@link BlobDescriptor}
+   * @throws LayerException if not available
+   */
+  public abstract BlobDescriptor getBlobDescriptor() throws LayerException;
+
+  /**
+   * @return the layer's diff ID
```
Comment (elharo) [inline crepecake/src/main/java/com/google/cloud/tools/crepecake/image/Layer.java]: this doesn't really tell me what a diff ID is though

File: crepecake/src/main/java/com/google/cloud/tools/crepecake/image/LayerException.java [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package com.google.cloud.tools.crepecake.image;
+
+/** {@link Exception} thrown when accessing non-existent properties of layers. */
```
Comment (elharo) [inline crepecake/src/main/java/com/google/cloud/tools/crepecake/image/LayerException.java]: no need to link to Exception

File: crepecake/src/main/java/com/google/cloud/tools/crepecake/image/LayerType.java [hunk 6]
```diff
@@ -0,0 +1,31 @@
+package com.google.cloud.tools.crepecake.image;
```
Comment (elharo) [inline crepecake/src/main/java/com/google/cloud/tools/crepecake/image/LayerType.java]: copyright

File: crepecake/src/main/java/com/google/cloud/tools/crepecake/image/CachedLayer.java (hunk 2 above)
Comment (coollog) [reply crepecake/src/main/java/com/google/cloud/tools/crepecake/image/CachedLayer.java]: When caching, the UnwrittenLayer is written out as its BLOB, descriptor, and diffId. There will be no checks at runtime in CachedLayer for matching and the user of CachedLayer must ensure that the descriptor and diffId matches the BLOB stored in the file. I will add a note about this in the javadoc.

File: crepecake/src/main/java/com/google/cloud/tools/crepecake/image/CachedLayer.java (hunk 3 above)
Comment (coollog) [reply crepecake/src/main/java/com/google/cloud/tools/crepecake/image/CachedLayer.java]: Yes, BlobDescriptor is immutable.

File: crepecake/src/main/java/com/google/cloud/tools/crepecake/image/Layer.java (hunk 4 above)
Comment (coollog) [reply crepecake/src/main/java/com/google/cloud/tools/crepecake/image/Layer.java]: I have the description above in the class javadoc - should I duplicate that to here as well?

File: crepecake/src/main/java/com/google/cloud/tools/crepecake/image/LayerType.java (hunk 6 above)
Comment (coollog) [reply crepecake/src/main/java/com/google/cloud/tools/crepecake/image/LayerType.java]: Added copyright.

File: crepecake/src/main/java/com/google/cloud/tools/crepecake/image/Layer.java [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+ *   <li>Diff ID
+ *   <li>
+ *       <ul>
+ *         <li>The SHA-256 hash of the uncompressed archive (tarball) of the partial filesystem
+ *             changeset.
+ *       </ul>
+ *
+ * </ul>
+ */
+public abstract class Layer {
```
Comment (loosebazooka) [inline crepecake/src/main/java/com/google/cloud/tools/crepecake/image/Layer.java]: Why does this need to be an abstract class?

Comment (loosebazooka) [reply crepecake/src/main/java/com/google/cloud/tools/crepecake/image/Layer.java]: and not an interface?

File: crepecake/src/main/java/com/google/cloud/tools/crepecake/image/LayerException.java [hunk 8]
```diff
@@ -0,0 +1,25 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package com.google.cloud.tools.crepecake.image;
+
+/** Exception thrown when accessing non-existent properties of layers. */
```
Comment (loosebazooka) [inline crepecake/src/main/java/com/google/cloud/tools/crepecake/image/LayerException.java]: Is this only thrown for Layer properties that don't exist? Can we call it something like that?

File: crepecake/src/main/java/com/google/cloud/tools/crepecake/image/CachedLayer.java [hunk 9]
```diff
@@ -0,0 +1,69 @@
...
+
+package com.google.cloud.tools.crepecake.image;
+
+import com.google.cloud.tools.crepecake.blob.BlobDescriptor;
+import com.google.cloud.tools.crepecake.blob.BlobStream;
+import com.google.cloud.tools.crepecake.blob.BlobStreams;
+import java.io.File;
+
+/**
+ * A {@link Layer} that has been written out (i.e. to a cache) and has its file-backed content BLOB,
```
Comment (elharo) [inline crepecake/src/main/java/com/google/cloud/tools/crepecake/image/CachedLayer.java]: written out to a cache

File: crepecake/src/main/java/com/google/cloud/tools/crepecake/image/CachedLayer.java (hunk 2 above)
Comment (elharo) [reply crepecake/src/main/java/com/google/cloud/tools/crepecake/image/CachedLayer.java]: Could the constructor be non-public to make it harder to create a layer where these pieces don't match? Or could we reorganize the API so they always match?

File: crepecake/src/main/java/com/google/cloud/tools/crepecake/image/Layer.java (hunk 4 above)
Comment (elharo) [reply crepecake/src/main/java/com/google/cloud/tools/crepecake/image/Layer.java]: No, that's good. Thanks.

File: crepecake/src/main/java/com/google/cloud/tools/crepecake/image/ReferenceLayer.java [hunk 10]
```diff
@@ -0,0 +1,53 @@
...
+public class ReferenceLayer extends Layer {
+
+  /** The {@link BlobDescriptor} of the compressed layer content. */
+  private final BlobDescriptor blobDescriptor;
+
+  /** The digest of the uncompressed layer content. */
+  private final DescriptorDigest diffId;
+
+  /** Instantiate with a {@link BlobDescriptor} and diff ID. */
+  public ReferenceLayer(BlobDescriptor blobDescriptor, DescriptorDigest diffId) {
```
Comment (elharo) [inline crepecake/src/main/java/com/google/cloud/tools/crepecake/image/ReferenceLayer.java]: Why is the DescriptorDigest independent of the BlobDescriptor instead of part of it? Could a single Blob have different descriptors?

Comment (coollog) [reply crepecake/src/main/java/com/google/cloud/tools/crepecake/image/ReferenceLayer.java]: A layer BLOB has its main DescriptorDigest that is a hash of the compressed layer tarball archive, but it also has a DescriptorDigest diff ID that is a hash of the uncompressed layer tarball archive.

File: crepecake/src/test/java/com/google/cloud/tools/crepecake/image/UnwrittenLayerTest.java [hunk 11]
```diff
@@ -0,0 +1,74 @@
...
+import org.mockito.Mockito;
+import org.mockito.MockitoAnnotations;
+
+/** Tests for {@link UnwrittenLayer}. */
+public class UnwrittenLayerTest {
+
+  @Rule public TemporaryFolder fakeFolder = new TemporaryFolder();
+
+  @Mock private Blob mockCompressedBlob;
+
```
Comment (elharo) [inline crepecake/src/test/java/com/google/cloud/tools/crepecake/image/UnwrittenLayerTest.java]: nit: probably no blank lines in between these Mock fields

**Comment by coollog** [comment]
@loosebazooka requesting a final pass

File: crepecake/src/main/java/com/google/cloud/tools/crepecake/blob/AbstractHashingBlob.java [hunk 12]
```diff
@@ -20,27 +20,28 @@
...
 
-    writeToAndHash(hashingOutputStream);
+    writeToWithHashing(hashingOutputStream);
     hashingOutputStream.flush();
 
-    return hashingOutputStream.toBlobDescriptor();
+    try {
+      return hashingOutputStream.toBlobDescriptor();
+    } catch (DigestException ex) {
+      throw new IOException("BLOB hashing failed: " + ex.getMessage());
```
Comment (elharo) [inline crepecake/src/main/java/com/google/cloud/tools/crepecake/blob/AbstractHashingBlob.java]: include the nested exception

File: crepecake/src/main/java/com/google/cloud/tools/crepecake/blob/AbstractHashingBlob.java [hunk 13]
```diff
@@ -20,27 +20,28 @@
...
-   * @param outputStream the {@link DigestOutputStream} to write to
+   * @param outputStream the {@link CountingDigestOutputStream} to write to
    */
-  protected abstract void writeToAndHash(OutputStream outputStream) throws IOException;
+  abstract void writeToWithHashing(CountingDigestOutputStream outputStream) throws IOException;
 
   @Override
-  public final BlobDescriptor writeTo(OutputStream outputStream)
-      throws IOException, DigestException {
+  public final BlobDescriptor writeTo(OutputStream outputStream) throws IOException {
```
[the remark this answers is not in the document]
Comment (coollog) [reply crepecake/src/main/java/com/google/cloud/tools/crepecake/blob/AbstractHashingBlob.java]: Okay, let me open a separate issue for that.
